src/services/FilesActionHandler.js

```diff
@@ -20,36 +20,31 @@
  *
  */
 
-import { encodePath } from '@nextcloud/paths'
-
 /**
- * @param {string} name the file name
- * @param {object} context the file context
+ * @param {Node} node The file to open
  */
-export default function(name, context) {
+export default function(node) {
 	// replace potential leading double slashes
-	const path = `${context.dir}/${name}`.replace(/^\/\//, '/')
-	const oldQuery = location.search.replace(/^\?/, '')
-	const onClose = () => OC.Util.History.pushState(oldQuery)
-	if (!context.fileInfoModel && context.fileList) {
-		context.fileInfoModel = context.fileList.getModelForFile(name)
-	}
-	if (context.fileInfoModel) {
-		pushToHistory({ fileid: context.fileInfoModel.get('id') })
-	}
+	const path = `${node.dirname}/${node.basename}`.replace(/^\/\//, '/')
+	const oldRoute = [
+		window.OCP.Files.Router.name,
+		window.OCP.Files.Router.params,
+		window.OCP.Files.Router.query,
+		true,
+	]
+	const onClose = () => window.OCP.Files.Router.goToRoute(...oldRoute)
+	pushToHistory(node)
 	OCA.Viewer.open({ path, onPrev: pushToHistory, onNext: pushToHistory, onClose })
 }
 
 /**
- * @param {object} root destructuring object
- * @param {number} root.fileid the opened file ID
+ * @param {Node} node Node to navigate to
  */
-function pushToHistory({ fileid }) {
-	const params = OC.Util.History.parseUrlQuery()
-	const dir = params.dir
-	delete params.dir
-	delete params.fileid
-	params.openfile = fileid
-	const query = 'dir=' + encodePath(dir) + '&' + OC.buildQueryString(params)
-	OC.Util.History.pushState(query)
+function pushToHistory(node) {
+	window.OCP.Files.Router.goToRoute(
+		null,
+		{ ...window.OCP.Files.Router.params, fileid: node.fileid },
+		{ dir: window.OCP.Files.Router.query.dir, openfile: node.fileid },
```

Comment:
A few comments:
- You have access to the current View from the action, so you can probably pass it here and use it directly.
- For the dir, you can use the file path directly
- And finally, openfile is not implemented yet, but it should probably just be set to true. Then we'll use the fileid param

```suggestion
		{ view: view.id, fileid: node.fileid },
		{ dir: node.dirname, openfile: true },
```

Reply:
Fixed, thanks for the hint